src/main/java/com/uid2/admin/vertx/service/AdminKeyService.java

```diff
@@ -211,112 +220,106 @@ private void handleAdminAdd(RoutingContext rc) {
 
             // respond with new admin created
             rc.response().end(jsonWriter.writeValueAsString(newAdmin));
+            return new OperationModel(Type.ADMIN, name, Actions.CREATE,
+                    DigestUtils.sha256Hex(jsonWriter.writeValueAsString(newAdmin)), "created " + name);
         } catch (Exception e) {
             rc.fail(500, e);
+            return null;
         }
     }
 
-    private void handleAdminDel(RoutingContext rc) {
+    private OperationModel handleAdminDel(RoutingContext rc) {
         try {
             // refresh manually
             adminUserProvider.loadContent(adminUserProvider.getMetadata());
 
-            final String name = rc.queryParam("name").get(0);
-            Optional<AdminUser> existingAdmin = this.adminUserProvider.getAll()
-                    .stream().filter(a -> a.getName().equals(name))
-                    .findFirst();
-            if (!existingAdmin.isPresent()) {
+            AdminUser a = getAdminUser(rc.queryParam("name"));
+            if (a == null) {
                 ResponseUtil.error(rc, 404, "admin not found");
-                return;
+                return null;
             }
 
             List<AdminUser> admins = this.adminUserProvider.getAll()
-                    .stream().sorted((a, b) -> (int) (a.getCreated() - b.getCreated()))
+                    .stream().sorted((x, y) -> (int) (x.getCreated() - y.getCreated()))
                     .collect(Collectors.toList());
 
             // delete admin from the array
-            AdminUser a = existingAdmin.get();
             admins.remove(a);
 
             // upload to storage
             storageManager.uploadAdminUsers(adminUserProvider, admins);
 
             // respond with admin deleted
             rc.response().end(jsonWriter.writeValueAsString(a));
+            return new OperationModel(Type.ADMIN, a.getName(), Actions.DELETE,
```

Comment:
Use email as key

Reply:
According to current code, only name is guaranteed to be unique